bauble/src/value/mod.rs

```diff
@@ -0,0 +1,723 @@
+use std::{borrow::Cow, collections::HashMap};
+
+use indexmap::IndexMap;
+use rust_decimal::Decimal;
+
+use crate::{
+    parse::{self, Object as ParseObject, Path, PathEnd, PathTreeEnd, Use, Values},
+    spanned::{SpanExt, Spanned},
+    VariantKind,
+};
+
+mod asset_context;
+
+pub use asset_context::*;
+
+#[derive(Clone)]
+pub struct Attributes(pub IndexMap<Spanned<String>, Val>);
+
+#[derive(Clone)]
+/// A value with attributes
+pub struct Val {
+    pub value: Spanned<Value>,
+    pub attributes: Spanned<Attributes>,
+}
+
+pub type Ident = Spanned<String>;
+
+pub type Map = Vec<(Val, Val)>;
+
+pub type Fields = IndexMap<Ident, Val>;
+
+pub type Sequence = Vec<Val>;
+
+// type path.
+type TypePath = TypeInfo;
+
+type AssetPath = String;
+
+#[derive(Clone)]
+pub enum FieldsKind {
+    Unit,
+    Tuple(Sequence),
+    Struct(Fields),
+}
+
+impl FieldsKind {
+    pub fn variant_kind(&self) -> VariantKind {
+        match self {
+            FieldsKind::Unit => VariantKind::Path,
+            FieldsKind::Tuple(_) => VariantKind::Tuple,
+            FieldsKind::Struct(_) => VariantKind::Struct,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ValueKind {
+    Unit,
+    Num,
+    Bool,
+    Str,
+    Opt,
+    Ref,
+    Array,
+    Map,
+    Tuple,
+    Struct,
+    Enum,
+    BitFlags,
+    Raw,
+}
+
+#[derive(Clone)]
+pub enum Value {
+    Unit,
+    Num(Decimal),
+    Bool(bool),
+    Str(String),
+    Opt(Option<Box<Val>>),
```

Comment:
There's no parser to create an `Opt`, which means `Option`s can't be deserialized. Though, I'd suggest getting rid of the `Opt` variant and flattening `Option`, like so:

```rust
// The Rust type
struct MyStruct {
    field: Option<f32>,
    other: Option<f32>,
}

// The `bauble` data
my_struct = MyStruct {
    field: 12.0,
    other: (), // Or perhaps `None`
}

// The resulting value
MyStruct {
    field: Some(12.0),
    other: None,
}
```

In my usage, all of my `Option` fields have `bauble(default)`, so I only use the fields in the markup if I have a value to give it. So, this design would help be less verbose, unless you prefer the verbosity anyway

Reply:
Oh, there doesn't seem to be a parser for `Unit` either. That's my bad, I think

Reply:
> There's no parser to create an Opt
The parser will create a `parse::value::Value::Tuple { name: "Some", .. }`, and when converted that will become an `Opt` in when resolving values. And similarly to None.